pyccel/codegen/printing/ccode.py

```diff
@@ -1466,16 +1466,92 @@ def _print_EmptyNode(self, expr):
     #=================== OMP ==================
     def _print_OMP_For_Loop(self, expr):
         omp_expr   = str(expr.txt)
-        return '#pragma omp for{}\n{{'.format(omp_expr)
+        return '#pragma omp for{}'.format(omp_expr)
 
     def _print_OMP_Parallel_Construct(self, expr):
-        omp_expr   = str(expr.txt)
-        return '#pragma omp {}\n{{'.format(omp_expr)
+        clauses = ''
+        if expr.combined:
+            clauses = ' ' + expr.combined
+        clauses += str(expr.txt)
+        omp_expr   = '#pragma omp parallel{}'.format(clauses)
+        if expr.combined is None:
+            omp_expr += '\n{'
+        elif (expr.combined and "for" not in expr.combined):
+            if "masked taskloop" not in expr.combined:
+                omp_expr += '\n{'
+        return omp_expr
 
     def _print_OMP_Single_Construct(self, expr):
         omp_expr   = str(expr.txt)
         return '#pragma omp {}\n{{'.format(omp_expr)
 
+    def _print_OMP_Critical_Construct(self, expr):
+        omp_expr = str(expr.txt)
+        omp_expr = '#pragma omp {}\n{{'.format(omp_expr)
+        return omp_expr
+
+    def _print_OMP_Master_Construct(self, expr):
+        omp_expr = str(expr.txt)
+        omp_expr = '#pragma omp {}\n{{'.format(omp_expr)
+        return omp_expr
+
+    def _print_OMP_Masked_Construct(self, expr):
+        clauses = ''
+        if expr.combined:
+            clauses = ' ' + expr.combined
+        clauses += str(expr.txt)
+        omp_expr = '#pragma omp masked{}'.format(clauses)
+        if expr.combined is None:
+            omp_expr += '\n{'
+        return omp_expr
+
+    def _print_OMP_Task_Construct(self, expr):
+        omp_expr = str(expr.txt)
+        omp_expr = '#pragma omp {}\n{{'.format(omp_expr)
+        return omp_expr
+
+    def _print_OMP_Target_Construct(self, expr):
+        clauses = ''
+        if expr.combined:
+            clauses = ' ' + expr.combined
+        clauses += str(expr.txt)
+        omp_expr = '#pragma omp target{}'.format(clauses)
+        if expr.combined is None:
+            omp_expr += '\n{'
+        elif (expr.combined and "for" not in expr.combined):
+            omp_expr += '\n{'
```

Comment:
Some of the combined cases make an exception for `for`, but not all. Is that intentional?

Reply:
because for example when we have `#$ omp parallel for` we know that were gonna have a for loop after the pragma so we don't need to have an `#$ omp end parallel for`

Reply:
Your answer seems to explain why you have `and "for" not in expr.combined`. My question wasn't why you have it here, but if it should be in every class with a `combined` property. E.g. [here](https://github.com/pyccel/pyccel/pull/593/files/56efb97a63fb0f4d99fe5c9bc6659bd9cb3afcfa#diff-f1038cb7408f08867011c5feb494cd9e7a5a435a3b6f6438ef6cd3bdf3c90fbeR1536)

In codegen/printing/ccode.py, sometimes you do:
```
elif (expr.combined and "for" not in expr.combined):
    omp_expr += '\n{'
```
and sometimes you do:
```
if expr.combined is None:
    omp_expr += '\n{
```
I'm not sure if this is due to a difference in the OMP statements or due to a correction that hasn't been done everywhere?

Reply:
but a `parallel` construct can be combined with the `sections` construct, so when it's combined with the `sections` construct we need to open the bracket

Reply:
it depend on whats inside the `combined` variable, I try to not open the bracket when there will be only one statement after the pragma

Reply:
The question is about the condition

Reply:
>it depend on whats inside the combined variable, I try to not open the bracket when there will be only one statement after the pragma

This is my point. Presumably all combined variables should avoid opening brackets if `"for" in expr.combined`? So why do some classes not have this condition?

Reply:
aah i see your question, because the `for` can't be combined with any construct, and if it's the first element after a construct we know that there is a `for` thank's to textx so we need to check just if `expr.combined is not None`

Reply:
like `#$ omp teams distribute`, here if there is any combined construct after the `teams` it needs to be a `distribute` construct, so we check it using textx if the `expr.combined is not None` then we don't open the bracket, and what come after the `distribute` is optional we need to check it

Reply:
Ok. So yes it is intentional :+1:

Reply:
Presumably it wouldn't matter if we did check if `"for" in expr.combined` even if it's not necessary? (Just wondering if this could help for reducing code duplication)

Reply:
but it could be `#$ omp parallel sections` this needs a structured block so we need to open the brackets

Reply:
Let me rephrase my 2nd question. Presumably doing:
```
def _print_OMP_Masked_Construct(self, expr):
        clauses = ''
        if expr.combined:
            clauses = ' ' + expr.combined
        clauses += str(expr.txt)
        omp_expr = '#pragma omp masked{}'.format(clauses)
        if expr.combined is None:
            omp_expr += '\n{'
        elif (expr.combined and "for" not in expr.combined):
            omp_expr += '\n{'
        return omp_expr
```
wouldn't cause any problems? I am asking as I am wondering if we could do:
```
def _print_OmpCombinedAnnotatedComment (self, expr):
        clauses = ''
        if expr.combined:
            clauses = ' ' + expr.combined
        clauses += str(expr.txt)
        omp_expr = '#pragma omp masked{}'.format(clauses)
        if expr.combined is None:
            omp_expr += '\n{'
        elif (expr.combined and "for" not in expr.combined):
            omp_expr += '\n{'
        return omp_expr
```

Reply:
aaah im gonna try to make a printer for the combined construct and manage the cases